tileserver/__init__.py

```diff
@@ -330,59 +365,69 @@ def reformat_from_stored_json(self, request_data, layer_data):
 
         # in any case, it makes sense to try and fetch the json format from
         # the store first
-        tile_data = self.read_tile(coord)
+        tile_data = self.read_tile(meta_coord, offset)
         if tile_data is None:
             return None
 
         # the json format exists in the store we'll use it to generate the
-        # response
-        tile_data = reformat_selected_layers(
-            tile_data, layer_data, coord, format, self.buffer_cfg)
+        # response. don't need to reformat if the tile is already in JSON.
+        if layer_spec != 'all' or format != json_format:
+            tile_data = reformat_selected_layers(
+                tile_data, layer_data, coord, format, self.buffer_cfg)
 
         if layer_spec == 'all':
             # for the all layer, since the json format existed, we should also
             # save the requested format too to allow the caches to serve it
             # directly in subsequent requests we'll guard against re-saving
             # json onto itself though, which may be possible through a race
             # condition
-            if format != json_format:
+            #
+            # we can't store if we're using metatiles, as the metatile should
+            # already contains all the formats, and might contain more than
+            # one coordinate!
+            if format != json_format and not self.using_metatiles():
```

Comment:
It shouldn't be used now, since we're `self.using_metatiles()` everywhere now. Should we keep it in case anyone else is using it, or just clean it up?